### PR TITLE
Add unit test project based on boost_unit_test_framework

### PR DESCRIPTION
**Pull request by yiylin**
- [ ] Does this PR have a descriptive title that could go in our release notes?
- [x] Does this PR add any new dependencies?
Add boost_unit_test_framework for the unit test project
- [ ] Does this PR modify any existing APIs?
   - [ ] Is the change to the API backwards compatible?
- [ ] Should this result in any changes to our documentation, either updating existing docs or adding new ones?
 
#### What does this implement/fix? Briefly explain your changes.
Add unit test project based on boost_unit_test_framework, so that devs could write the real unit tests.

### CODE REVIEW

**Comment by daxpryce** [comment]
I have no preference to unit test library - I believe @gopalrs was advocating for googletest before, but I really don't know the libraries.

I would like to see some example of the actual unit test happening - maybe you could do one around the IndexWriteParameters in include/parameters.h?  It's super trivial with only a few caveats (like get_omp_max_threads() being exercised in case threads is set to 0, for instance?)

Just hoping to see what the `assertTrue` type stuff is like :)

**Comment by yiylin** [comment]
> I have no preference to unit test library - I believe @gopalrs was advocating for googletest before, but I really don't know the libraries.
> 
> I would like to see some example of the actual unit test happening - maybe you could do one around the IndexWriteParameters in include/parameters.h? It's super trivial with only a few caveats (like get_omp_max_threads() being exercised in case threads is set to 0, for instance?)
> 
> Just hoping to see what the `assertTrue` type stuff is like :)

Added the example

**Review by gopalrs — COMMENTED** [review]
I don't know enough about the YAML file or BOOST_TEST to do a detailed review, but at a high level this looks OK. 
Can we add a pointer on how to write new test cases somewhere in the documentation? (Can simply be a link to a site that describes the process of adding new test cases)

**Comment by yiylin** [comment]
> I don't know enough about the YAML file or BOOST_TEST to do a detailed review, but at a high level this looks OK. Can we add a pointer on how to write new test cases somewhere in the documentation? (Can simply be a link to a site that describes the process of adding new test cases)

Add a README.md inside the tests folder

File: .github/workflows/unit-tests.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      - name: Checkout repository
+        if: ${{ runner.os == 'Windows' }}
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 1
+          submodules: true
+      - name: DiskANN Build CLI Applications
+        uses: ./.github/actions/build   
+        
+      - name: Run Unit Tests
```
Comment (yiylin) [inline .github/workflows/unit-tests.yml:29]: image

Comment (yiylin) [reply .github/workflows/unit-tests.yml:29]: this shows that the unit tests are ran. I only add one test case, so here only shows 1 test

Comment (yiylin) [reply .github/workflows/unit-tests.yml:29]: this shows that the unit test project is ran. I only add one test case, so here only shows 1 test

**Comment by yiylin** [comment]
> > I have no preference to unit test library - I believe @gopalrs was advocating for googletest before, but I really don't know the libraries.
> > I would like to see some example of the actual unit test happening - maybe you could do one around the IndexWriteParameters in include/parameters.h? It's super trivial with only a few caveats (like get_omp_max_threads() being exercised in case threads is set to 0, for instance?)
> > Just hoping to see what the `assertTrue` type stuff is like :)
> 
> Added the example

Gopal is fine to use BOOST TEST

File: .github/actions/build/action.yml [hunk 2]
```diff
@@ -23,6 +23,6 @@ runs:
         mkdir build && cd build && cmake .. && msbuild diskann.sln /m /nologo /t:Build /p:Configuration="Release" /property:Platform="x64" -consoleloggerparameters:"ErrorsOnly;Summary"
         cd ..
         mkdir dist
-        mv .\x64\Release\ .\dist\bin
+        mklink /j .\dist\bin .\x64\Release\
```
Comment (daxpryce) [inline .github/actions/build/action.yml:26]: what does this do?  is this why I couldn't get the exes to run from a different directory?  teach me! :D

**Review by daxpryce — APPROVED** [review]
looks good to me!

File: Dockerfile [hunk 3]
```diff
@@ -7,7 +7,7 @@ RUN apt update
 RUN apt install -y software-properties-common
 RUN add-apt-repository -y ppa:git-core/ppa
 RUN apt update
-RUN DEBIAN_FRONTEND=noninteractive apt install -y git make cmake g++ libaio-dev libgoogle-perftools-dev libunwind-dev clang-format libboost-dev libboost-program-options-dev libmkl-full-dev libcpprest-dev python3.10
+RUN DEBIAN_FRONTEND=noninteractive apt install -y git make cmake g++ libaio-dev libgoogle-perftools-dev libunwind-dev clang-format libboost-dev libboost-program-options-dev libboost-test-dev libmkl-full-dev libcpprest-dev python3.10
```
[the remark this answers is not in the document]
Comment (daxpryce) [reply Dockerfile:10]: oh, to make it even more fun, the python musllinux wheel we're going to build needs to build on alpine linux, which uses apk and differenterer package names

[the remark this answers is not in the document]
Comment (yiylin) [reply Dockerfile:10]: sure. we already have this dependency "boost-program-options", and I expect that the naming for boost-test should be similar with "boost-program-options"

[the remark this answers is not in the document]
Comment (yiylin) [reply Dockerfile:10]: and this package is only for the unit test project. if you do not need to run the unit test project, we also do not need to have this dependency in your case

File: tests/index_write_parameters_builder_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
+#include <boost/test/unit_test.hpp>
```
Comment (harsha-simhadri) [inline tests/index_write_parameters_builder_tests.cpp:1]: add standard header comments

File: tests/index_write_parameters_builder_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+#include "parameters.h"
+
+BOOST_AUTO_TEST_SUITE(IndexWriteParametersBuilder_tests)
+
+BOOST_AUTO_TEST_CASE(test_build)
+{
+    uint32_t search_list_size = rand();
```
[the remark this answers is not in the document]
Comment (yiylin) [reply tests/index_write_parameters_builder_tests.cpp:9]: this is just a unit test to check if the value in the final parameters is expected and we do not have this checking in the Builder class. so far, I think it is fine to keep it like this. we do not use this parameter to build graph. If you think we can not set too high value, then we need to add some validation logic in the builder class.

File: Dockerfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (yiylin) [reply Dockerfile]: As Harsha suggests, I add a cmake flag "UNIT_TEST" to enable the unit test and add a new Dockerfile(DockerfileDev) for developers.

**Pull request merged**